File: ui/lib/client/consumerStats.dart
class ConsumerStats {
  ConsumerStats(
      this.id,
      this.totalRecords,
      this.recentRecords,
      this.errors
      );

  String id;
  int totalRecords;
  List<RecordSummary> recentRecords = [];
  List<RecordSummary> errors = [];

  @override String toString() => asJson.toString();

  Map<String, Object> get asJson {
    return {
      'id': id,
      'totalRecords': totalRecords,
      'recentRecords': recentRecords,
      'errors': errors
    };
  }

  static ConsumerStats fromJson(Map<String, dynamic> json) {
    final List<dynamic> records = json['recentRecords'];
    final List<dynamic> errors = json['errors'];
    return ConsumerStats(
        json['id'],
        json['totalRecords'],
        records.map((j) => RecordSummary.fromJson(j)).toList(),
        errors.map((j) => RecordSummary.fromJson(j)).toList());
  }
}
class RecordSummary {
  RecordSummary(
      this.record,
      this.message,
      this.value,
      this.timestampEpochMillis,
      this.supplementaryData
      );

  RecordCoords record;
  String message;
  dynamic value;
  int timestampEpochMillis;
  dynamic supplementaryData;

  Map<String, Object> get asJson {
    return {
      'record': record,
      'message': message,
      'value': value,
      'timestampEpochMillis': timestampEpochMillis,
      'supplementaryData': supplementaryData
    };
  }

  static RecordSummary fromJson(Map<String, dynamic> json) {
    return RecordSummary(
        RecordCoords.fromJson(json['record']),
        json['message'],
        json['value'],
        json['timestampEpochMillis'],
        json['supplementaryData']);
  }
}

class RecordCoords {
  RecordCoords(
      this.topic,
      this.offset,
      this.partition,
      this.key
      );

  String topic;
  int offset;
  int partition;
  String key;

  Map<String, Object> get asJson {
    return {
      'topic': topic,
      'offset': offset,
      'partition': partition,
      'key': key
    };
  }

  static RecordCoords fromJson(Map<String, dynamic> json) {
    return RecordCoords(
        json['topic'],
        json['offset'],
        json['partition'],
        json['key']);
  }
}

File: client/jvm/src/main/scala/expressions/client/kafka/ConsumerStats.scala
package expressions.client.kafka

import io.circe.Json
import io.circe.Codec

case class RecordCoords(topic: String, offset: Long, partition: Int, key: String)

object RecordCoords {
  given codec : Codec[RecordCoords] = io.circe.generic.semiauto.deriveCodec[RecordCoords]
}

case class RecordSummary(record: RecordCoords, message: String, value: Json, timestampEpochMillis: Long, supplementaryData: Json = Json.Null)

object RecordSummary {
  given codec : Codec[RecordSummary] = io.circe.generic.semiauto.deriveCodec[RecordSummary]
}

case class ConsumerStats(id: String,
                         totalRecords: Long = 0L,
                         stdout :Seq[String] = Nil,
                         stderr :Seq[String] = Nil,
                         recentRecords: Seq[RecordSummary] = Nil,
                         errors: Seq[RecordSummary] = Nil) {
  def ++(batch: Seq[RecordSummary]): ConsumerStats = {
    copy(recentRecords = (batch ++ recentRecords).take(100), totalRecords = totalRecords + batch.size)
  }
}

object ConsumerStats {
  given codec : Codec[ConsumerStats] = io.circe.generic.semiauto.deriveCodec[ConsumerStats]
}
